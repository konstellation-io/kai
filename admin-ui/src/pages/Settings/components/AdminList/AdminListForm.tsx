import React, { ReactElement, useEffect } from 'react';
import { get } from 'lodash';
import styles from './AdminListForm.module.scss';
import ItemList from '../../../../components/ItemList/ItemList';
import FormField from '../FormField/FormField';
import { useForm } from 'react-hook-form';

type OnValidateType = (...args: any) => string | boolean;

type AdminListProps = {
  onSubmit: Function;
  onRemoveItem: Function;
  itemName: string;
  itemList: string[];
  error?: string;
  isLoading?: boolean;
  onValidate?: OnValidateType;
  icon?: ReactElement;
};

function AdminListForm({
  onSubmit,
  onRemoveItem,
  itemName,
  itemList,
  error,
  isLoading = false,
  onValidate = () => true,
  icon = <></>
}: AdminListProps) {
  const {
    handleSubmit,
    setValue,
    register,
    errors,
    watch,
    clearError
  } = useForm();

  useEffect(() => {
    register('item', {
      validate: value => onValidate(value, itemList)
    });
    setValue('item', '');
  }, [register, setValue, itemList, onValidate]);

  function handleOnSubmit(formData: any) {
    onSubmit(formData);
    setValue('item', '');
  }

  function handleRemoveItem(args: any) {
    setValue('item', '');
    clearError('item');
    onRemoveItem(args);
  }

  return (
    <div className={styles.container}>
      <FormField
        fieldName={itemName}
        error={get(errors.item, 'message')}
        onChange={(value: string) => setValue('item', value)}
        onSubmit={handleSubmit(handleOnSubmit)}
        value={watch('item', '')}
        icon={icon}
      />
      <div className={styles.items}>
        <ItemList
          onRemoveItem={handleRemoveItem}
          error={error}
          loading={isLoading}
          data={itemList}
        />
      </div>
    </div>
  );
}

export default AdminListForm;
