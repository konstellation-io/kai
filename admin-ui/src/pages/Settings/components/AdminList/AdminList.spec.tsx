import React from 'react';
import { shallow } from 'enzyme';
import AdminListForm from './AdminListForm';

describe('AdminListForm component', () => {
  let wrapper: any;

  const onSubmitMock = jest.fn();
  const itemListMock = ['foo', 'bar', 'baz', 'qux'];
  const onRemoveItemMock = jest.fn();
  const itemNameMock = 'FOO';
  const errorMock = 'error';
  const onValidateMock = jest.fn();
  const isLoadingMock = false;

  beforeEach(() => {
    jest.resetAllMocks();
    wrapper = shallow(
      <AdminListForm
        onSubmit={onSubmitMock}
        itemList={itemListMock}
        onRemoveItem={onRemoveItemMock}
        itemName={itemNameMock}
        error={errorMock}
        onValidate={onValidateMock}
        isLoading={isLoadingMock}
      />
    );
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
