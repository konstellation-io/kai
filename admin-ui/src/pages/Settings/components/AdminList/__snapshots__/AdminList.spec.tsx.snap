// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminListForm component should render without crashing 1`] = `
<div
  className="container"
>
  <FormField
    fieldName="FOO"
    icon={<React.Fragment />}
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
  <div
    className="items"
  >
    <ItemList
      data={
        Array [
          "foo",
          "bar",
          "baz",
          "qux",
        ]
      }
      error="error"
      loading={false}
      onRemoveItem={[Function]}
    />
  </div>
</div>
`;
